A Release Schedule and Version Numbering Scheme for SunPy (#40)
sunpy/sunpy-SEP: SEP-0009.md

| title         | Release Pattern for SunPy Core |
| author(s)     | Stuart Mumford, Russell Hewett |
| contact email | stuart@cadair.com              |
| date-creation | 2018-12-13                     |
| type          | process                        |
| status        | discussion                     |
| discussion    |                                |

# Introduction

This SEP defines a release pattern for the SunPy core package and
defines a versioning system. Since SunPy is being used more widely both by
users as well as developers as a dependency, it is important to have a clear
and predictable releases.  This is especially important to support other
Python packages which have sunpy as a dependency (e.g. affiliated packages).
A predictable release and support cycle for the SunPy core package will aid
these packages in planning their own development cycle.

The focus of this SEP is not to prescribe how quickly the core package improves
through the introduction of new features or API changes but how and when
release occur and how they are described.

# Detailed Description

## Release Schedule

This SEP prescribes that there will be at least two releases of the sunpy core

> package per year with at least 4 months (preferably 6 months) between these

Reviewer: This could mean that we could have up to an infinite number of releases in a year so long as there are at least four months between the first and last releases.  Is this what is meant?  If not, is this a loophole that we don't want to exploit in the future?  Does this need to be clarified?

Author: I think I have clarified this in the newest commit, can you take a look?

releases. It is suggested that these occur in the months of May and
November based on the current release schedule of the astropy core package. It
is the responsibility of the lead developer to create a release schedule that
is consistent with these requirement and to present it to the Board at the
beginning of every year.  Releases are recommended to be timed to pertinent

> release cycles of upstream packages.

Reviewer 2: And that would be published on the wiki? on the docs? as a new project/PR?
I would prefer we say here where to expect that information, and how long the beginning of the year would take... is that the first week of January? is that before end of February?

Author: Well I would imagine it in reality wont change very often, so I don't think timing is that important. To be honest I think the best canonical reference for the release schedule is this SEP?

An example release sequence, following this scheme, would look like this:

* 1.0.0 is an LTS release, released November 2019.
* 1.0.1 is a bugfix to the LTS release, released December 2019.
* 1.1.0 is a short support release, released May 2020.
* 1.0.2 and 1.1.1 are bugfix releases, released June 2020, to fix a current bug and backport the fix to the LTS version.
* 1.2.0 is an unplanned short support release, released August 2020
* 2.0.0 is an LTS release, released November 2020.

Every numbered release shall be associated with a Digital Object Identifier
(DOI) and it is recommended that major LTS release be supported by journal
article if appropriate.

## Deprecations and Documentation

The SunPy core package must always document API changes, in the CHANGELOG and in
the user documentation. Where practical a side-by-side comparison of old and new
functionality will be provided.

In addition to this, where possible, code will emit deprecation warnings
to inform users of planned changes to or removal of API. Warnings will **only be
emitted** when an alternative option exists for the functionality, or the
functionality is to be completely removed.

> All effort will be made to provide at least one release (LTS or short suppoer)

Reviewer 2: ```suggestion
All effort will be made to provide at least one release (LTS or short support)
```

# SEP-0009 -- Release Pattern for SunPy Core

| SEP           | 9                              |
|---------------|--------------------------------|
| title         | Release Pattern for SunPy Core |
| author(s)     | Stuart Mumford, Russell Hewett |
| contact email | stuart@cadair.com              |
| date-creation | 2018-12-13                     |
| type          | process                        |
| status        | discussion                     |
| discussion    |                                |

# Introduction

This SEP defines a release pattern for the SunPy core package and
defines a versioning system. Since SunPy is being used more widely both by

> users as well as developers as a dependency, it is important to have a clear

Reviewer 3: ```suggestion
users as well as developers as a dependency, it is important to have clear
```

and predictable releases.  This is especially important to support other

> Python packages which have sunpy as a dependency (e.g. affiliated packages).

Reviewer 3: ```suggestion
Python packages which have SunPy as a dependency (e.g. affiliated packages).
```

A predictable release and support cycle for the SunPy core package will aid
these packages in planning their own development cycle.

The focus of this SEP is not to prescribe how quickly the core package improves
through the introduction of new features or API changes but how and when
release occur and how they are described.

# Detailed Description

## Release Schedule

> This SEP prescribes that there will be two planned releases of the sunpy core

Reviewer 3: ```suggestion
This SEP prescribes that there will be two planned releases of the SunPy core
```

package per year with at least 4 months (preferably 6 months) between these
releases. It is suggested that these occur in the months of May and November
based on the current release schedule of the astropy core package. It is the
responsibility of the lead developer to plan a release schedule that is
consistent with these requirement and to present it to the Board at the
beginning of every year, if it differs from this recommendation. Releases are
recommended to be timed to pertinent release cycles of upstream packages. The
lead developer can at their discretion do an out of cycle release if they feel
there is sufficient justification.

## Supporting Releases with bug fixes

This SEP defines two categories of SunPy release, a Long

> Term Support (LTS) release and a short-support or non-LTS release. Both types of

Reviewer 4: short support sounds weird. Call these normal releases, or feature releases or just simply release?

releases will be supported by bug fixes and documentation upgrades but for
differing amounts of time. The first release
of the year shall be a non-LTS release, which shall be supported
for 6 months or until the next release. The second release of the year
will be an LTS release and will be supported for 12 months or until
the next LTS release. Support periods can be extended beyond the requirements
here if needed.

The objective of this support cycle is to balance the desire to provide a

> release cycle for people who wish to experience minimal changes to SunPy with

Reviewer 3: ```suggestion
release cycle for people who wish to experience minimal changes to SunPy while
```

minimizing the maintenance burden on the SunPy developers. This release schedule
is consistent with the resources available at the time of writing. Should more
(or less) resources become available this topic must be re-evaluated.

## Release Numbering

The following versioning system shall be used.

X.Y.z

Where the three components have the following meanings:

* X. is the LTS version number, this will be incremented with every LTS.
* Y. is the release counter, this will be 0 for LTS releases and increment for each intermediate release.
* z. is the bug fix number, and is to be incremented for any bug fix releases.

An example release sequence, following this scheme, would look like this:

* 1.0.0 is an LTS release, released November 2019.
* 1.0.1 is a bugfix to the LTS release, released December 2019.
* 1.1.0 is a short support release, released May 2020.
* 1.0.2 and 1.1.1 are bugfix releases, released June 2020, to fix a current bug and backport the fix to the LTS version.
* 1.2.0 is an unplanned short support release, released August 2020
* 2.0.0 is an LTS release, released November 2020.

Every numbered release shall be associated with a Digital Object Identifier

> (DOI) and it is recommended that major LTS release be supported by journal

Reviewer 3: ```suggestion
(DOI) and it is recommended that major LTS releases be supported by a journal
```

article if appropriate.

## Deprecations and Documentation

The SunPy core package must always document API changes, in the CHANGELOG and in
the user documentation. Where practical a side-by-side comparison of old and new
functionality will be provided.

In addition to this, where possible, code will emit deprecation warnings
to inform users of planned changes to or removal of API. Warnings will **only be
emitted** when an alternative option exists for the functionality, or the
functionality is to be completely removed.

> All effort will be made to provide at least one LTS release where functionality

Author: This change came out of a conversation with @dpshelio 

@nabobalis what's your thoughts on this?

Reviewer 5: I'd ask someone smarter.

Reviewer 4: I don't totally get what you mean in the paragraph, basically you pledge to get at least a 2 year deprecation cycle?

Reviewer 3: Shall we have a time commitment here instead of "one LTS release"? In case the frequency of LTS releases changes...

Author: The purpose of this is to ensure that if any users go from LTS to LTS release they still get a deprecation warning about pending changes. So with the planned release cadence it would be a one year deprecation cycle.

@tiagopereira What do you see as the value in a time based deprecation policy? I think that if the cadence of the LTS releases change then the deprecation time would also want to change along with it, so that as I mentioned people going from LTS to LTS will always get a warning.

Reviewer 3: I'm not sure if a time-based policy is the way to go. I just wanted to ensure that API changes do not happen very often, in case LTS releases become very frequent. Probably fine to keep the current wording.

Author: The idea of this SEP isn't to overly restrict the ability for us to make API changes, but the idea of an LTS is that you get at least a year of support on that release, so you can opt into slow API changes by sticking on that LTS and then moving onto whatever the current LTS is after a year.

Reviewer 4: >LTS releases become very frequent

By definition they cannot be that frequent, from the practical point of view sunpy is manure enough that no one wants to support multiple releases for years.

@Cadair - I missed the point that LTSes are proposed to be a yearly thing, a minimum ~1-year deprecation seems totally reasonable then.

and packages in the wider solar or heliophysics ecosystems. A predictable
release and support cycle for the SunPy core package will aid these packages in
planning their own development cycle.

The desire for stability and predictability is countered by the desire of the
SunPy developers to continue providing new features and functionality and
continue to iterate on the SunPy API to further increase our usability and
correctness.

This SEP therefore tries to focus on when the SunPy core package will be
released rather than placing any strict limitations on changes and API
stability.

# Detailed Description

## Release Schedule

This SEP proposes that the SunPy core package will be released at least twice a
year, initially a release in the months of May and November have been chosen to
place the SunPy release shortly after that of the Astropy core package. The
decision to make a release and the specific timing of releases are at the
discretion of the lead developer, and changes to the months of the releases can
be made without modifying this SEP to follow any pertinent release cycles of
upstream packages or any other reason.

> Future changes to the LTS release cycle should be informed by upstream package

Reviewer 6: need to define LTS.

> development. It is important that upstream packages versions are supported for

Reviewer 6: Not sure what you are trying to say with this last sentence.

Author: I mean if SunPy 0.9 depends on Astropy 3.1 then we can't support 0.9 longer than Astropy 3.1 is supported.

the lifetime of a SunPy LTS.

## Supporting Releases with bug fixes

This SEP proposes that there be two categories of SunPy release, a LTS (Long
Term Support) release and a normal or non-LTS release. The LTS release will have

> an extended length of bug fixes provided by the SunPy developers. We propose

Reviewer 6: what does "extended length of bug fixes" mean

Author: I mean that we could provide bug fixes for longer. I.e. support that release for 18 months rather than 12.

that the first release in a year be a non-LTS release, which will be supported
for 6 months (until the next release), the second release in a year will be an

> LTS release and will be supported for 12 months (until the next LTS release).

Reviewer 6: say that 12 months is the minimum.

Author: I would rather keep this concrete at 12, but maybe add some comment about the ED being able to change it?

The objective of this support cycle is to balance the desire to provide a
release cycle for people who wish to experience minimal changes to SunPy with

> minimizing the maintenance burden on the SunPy developers. It is feasible that

Reviewer 6: Suggest to remove sentence starting from "It is feasible..." to the end of paragraph. May not age well.

we could extend the support cycle of the LTS release to longer than 12 months,
to give users an overlap period where they can upgrade. However, at this time,
supporting more than two releases simultaneously is beyond the
resources of the project.

## Release Numbering

In the numbering of SunPy releases we wish to covey if the release is an LTS
release or not. We therefore propose the following versioning system.

X.Y.z

Where the three components have the following meanings:

* X. is the LTS version number, this will be incremented with every LTS.
* Y. is the release counter, this will be 0 for LTS releases and increment for each intermediate release.
* z. is the bug fix number, and is to be incremented for any bug fix releases.

A typical release sequence, following this scheme, would look like this:

* 1.0.0 is an LTS release, released November 2019.
* 1.0.1 is a bugfix to the LTS release, released December 2019.

> * 1.1.0 is a short support release, released May 2020.

Reviewer 6: 1.1.0 includes new features? or just bug fixes?

Author: new features.

* 1.0.2 and 1.1.1 are bugfix releases, released June 2020, to fix a current bug and backport the fix to the LTS version.
* 1.2.0 is an unplanned short support release, released August 2020
* 2.0.0 is an LTS release, released November 2020.
* And so on...

> ## Deprecations and Documentation

Reviewer 6: what is the relationship between version numbers and deprecation warnings?

Author: Only that we should deprecate functionality for *at least* one release before we remove it normally.
